Implement set in terms of dict

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117631
* #117625
* #117639
* #108420
* __->__ #110524
* #117630
* #112252

This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

lezcano committed:
Implement set in terms of dict

This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano commented:
The failure is unrelated

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

jansel commented:
Feel free to re-request review once tests are passing in the lower down PRs are fixed.

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano commented:
@jansel this PR and the next one should be ready for review as well

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -907,6 +891,28 @@ def call_custom_dict(tx, user_cls, *args, **kwargs):
+                ListIteratorVariable,
+            ),
+        ):
+            items = set(arg.unpack_var_sequence(tx))
```
True and true. That `set` call was unnecessary anyway, removed.

jansel commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1774,6 +1772,11 @@ def __call__(self, tx, value) -> VariableTracker:
             return UserDefinedClassVariable(value)
+        elif isinstance(value, set):
+            return SetVariable(
+                {ConstantVariable.create(x) for x in value},
```
Similar to other comments.  Usage of Python sets.

Since `value` is already a set, there will be some nondeterminism here.  Perhaps add a comment warning about that, since we might want to try to fix it later.

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -898,6 +882,28 @@ def call_custom_dict(tx, user_cls, *args, **kwargs):
+        # Can we merge this implementation and call_dict's one?
+        assert not kwargs
+        if not args:
+            return SetVariable({}, mutable_local=MutableLocal())
```
Don't use a `{}` here for consistent typing.

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano commented:
Fixed those two and audited the other calls to `SetVariable`'s constructor to make sure everything was a list (there was one that was built from `dict.keys()`).

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Implement set in terms of dict"


This allows to heavily simplify the implementation of set, which was
"quite unique". Now we represent a set a as a dict where all its values
are None.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
